### ISSUE TITLE
Command line parameters should follow conventions

### ISSUE THREAD

**Issue by patriksvensson** [cake-build/cake#474]
We should add support for the `--longoption=value` convention and (sometime in the future) deprecate the (non conventional) `-longoption=value`.

**Comment by gep13** [comment]
Why deprecate?  Why not support both?

**Comment by patriksvensson** [comment]
@gep13 Because it's wrong :smile:

Jokes aside, we want Cake to be intuitive, and for a *nix person this is confusing. Multiple letters combined with a single dash is multiple switches in their short form.

I.e. `-test` is the same as `-t -e -s -t`.

**Comment by gep13** [comment]
Ah, I see, that makes sense :+1:

**Comment by DevWouter** [comment]
Let's see if I can get this to work :)

**Comment by DevWouter** [comment]
Alright, since I haven't updated the issue with some details, here they are.

This is mostly about the parser.

First a few valid arguments:
```ps
# Long keys are always two dashes followed by at least two characters
.\cake --longkey value
# Short keys are always dash followed by a single char
.\cake -s short-key
# Duplicate keys are supported by the parser (might not be supported by the data-structure)
# In the following example the "keys" property should be a list.
.\cake --keys a --keys b
# Supports quotes
.\cake --double-quote "here we a space"
.\cake --single-quote 'here we a space'
# Supports escaped quotes
.\cake --double-quote "write a \"quote\""
.\cake --single-quote 'write a \'quote\''
# Values are not required
.\cake --no-value --one-with-value a_value --yet-no-value-again
```

And one of the last thing I have been thinking of is using list seperator like git uses. However unles there is use-case for it we might rather avoid it.
```powershell
.\cake --k1 --k2 -- And here is a long argument which can be anythig
```

**Comment by DevWouter** [comment]
From the chat with @patriksvensson 

**tl;dr:** Multiple arguments are supported by the parser but can't be used since `ICakeArguments` can't support it. 

```
patriksvensson 13:59
Well, I'm not against supporting multiple keys, but that would break the ICakeArguments interface. That's my main concern.

DevWouter 14:01
@patriksvensson Yes, but what the parsers supports can be more than the data-structure.
@patriksvensson We will still throw an error if that happens, but it won't be thrown by the parser as the grammar is correct.
```

**Comment by DevWouter** [comment]
Reading through the ticket I noticed that the flags aren't supported yet. It should be quite simple to add them now.  First I want to collect some feedback on the PR (since it is my first) and then see what else requires work so we can close this ticket.

**Comment by gep13** [comment]
**NOTE:** Once this has been completed, we need to update the documentation here:

http://cakebuild.net/docs/cli/usage

**Comment by reicheltp** [comment]
Take a look at https://github.com/dotnet/corefxlab/tree/master/src/System.CommandLine
This should be the easiest way to get this done.

**Comment by patriksvensson** [comment]
@reicheltp Thanks for the suggestion. The solution to this is just a couple of lines of code to the existing parser so I don't think taking another dependency is necessary.
